cmd/relay-s3/main.go

```diff
@@ -0,0 +1,194 @@
+		if cctx.IsSet("metrics") {
+			metricsListener := metrics.Start(cctx.String("metrics"), pprof.WithProfile(), nil)
+			defer metricsListener.Close()
+		}
+
+		var err error
+		var chainHash []byte
+
+		if cctx.String("chain-hash") != "" {
+			chainHash, err = hex.DecodeString(cctx.String("chain-hash"))
+			if err != nil {
+				return fmt.Errorf("decoding chain hash: %w", err)
+			}
+		}
+
+		sess, err := session.NewSession(&aws.Config{Region: aws.String(cctx.String("region"))})
+		if err != nil {
+			return fmt.Errorf("creating aws session: %w", err)
+		}
+
+		_, err = sess.Config.Credentials.Get()
+		if err != nil {
+			return fmt.Errorf("checking credentials: %w", err)
+		}
+
+		uploader := s3manager.NewUploader(sess)
+
+		var w client.Watcher
+
+		if len(cctx.StringSlice("http-connect")) > 0 {
+			w, err = newHTTPWatcher(cctx.StringSlice("http-connect"), chainHash)
+		} else if cctx.String("grpc-connect") != "" {
+			w, err = newGRPCWatcher(cctx.String("grpc-connect"), cctx.String("cert"), cctx.Bool("insecure"))
+		} else {
+			return errors.New("missing GRPC or HTTP address(es)")
+		}
+		if err != nil {
+			return err
+		}
+
+		ctx := context.Background()
+		for res := range w.Watch(ctx) {
+			log.Info("relay_s3", "got randomness", "round", res.Round())
+			go uploadRandomness(ctx, uploader, cctx.String("bucket"), res)
+		}
+		return nil
+	},
+}
+
+func newGRPCWatcher(addr string, cert string, insecure bool) (client.Watcher, error) {
+	return nil, errors.New("not implemented")
+}
+
+func newHTTPWatcher(urls []string, chainHash []byte) (client.Watcher, error) {
+	if chainHash == nil {
+		return client.New(client.WithInsecureHTTPEndpoints(urls), client.WithCacheSize(0))
+	}
+	return client.New(client.WithChainHash(chainHash), client.WithHTTPEndpoints(urls), client.WithCacheSize(0))
+}
+
+func uploadRandomness(ctx context.Context, uploader *s3manager.Uploader, bucket string, result client.Result) {
+	rd, ok := result.(*client.RandomData)
+	if !ok {
+		log.Error("relay_s3", "unexpected underlying result type")
+		return
+	}
+	data, err := json.Marshal(rd)
+	if err != nil {
+		log.Error("relay_s3", "failed to marshal randomness", "error", err)
+		return
+	}
+	for {
+		r, err := uploader.UploadWithContext(ctx, &s3manager.UploadInput{
+			ACL:          aws.String("public-read"),
+			Bucket:       aws.String(bucket),
+			Key:          aws.String(fmt.Sprintf("public/%v", result.Round())),
+			Body:         bytes.NewBuffer(data),
+			ContentType:  aws.String("application/json"),
+			CacheControl: aws.String("immutable"),
+		})
+		if err != nil {
+			log.Error("relay_s3", "failed to upload randomness", "round", result.Round(), "error", err)
+			t := time.NewTimer(uploadRetryInterval)
+			select {
+			case <-t.C:
+				continue
+			case <-ctx.Done():
+				t.Stop()
+				return
+			}
+		}
+		log.Info("relay_s3", "uploaded randomness", "round", result.Round(), "location", r.Location)
+		break
+	}
+	// TODO: upload to /public/latest
```

Comment:
i think it's fine to not maintain this for the s3 variant. clients can poll on the expected /public/n, or another relay layer can use this for historic polls, and having to maintain a mutable endpoint increases complexity for caching in a way that may not be worth it.

Reply:
yeah that would simplify things